File: solutions/2_Add_Two_Numbers/2_add_two_numbers_linked_list.cpp
/**
 * author: yfzhang
 * date: 2017-03-07
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* res = new ListNode(-1);
        ListNode* pres = res;
        int carry = 0;
        for (ListNode* p1 = l1, *p2 = l2; p1 != NULL || p2 != NULL; p1 = (p1 == NULL ? NULL : p1->next), p2 = (p2 == NULL ? NULL : p2->next), pres = pres->next)
        {
            int a1 = p1 == NULL? 0:p1->val;
            int a2 = p2 == NULL? 0:p2->val;
            int val = a1 + a2 + carry;
            carry = val / 10;
            val = val % 10;
            pres->next = new ListNode(val);
        }
        if (carry > 0)
        {
            pres->next = new ListNode(carry);
        }
        return res->next;
    }
};

File: solutions/2_Add_Two_Numbers/2_add_two_numbers_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param {ListNode} l1
    # @param {ListNode} l2
    # @return {ListNode}
    def addTwoNumbers(self, l1, l2):
        carry = 0
        ret_l = ListNode(0)
        tmp_l = ret_l
        while l1 and l2:
            val = l1.val + l2.val + carry
            tmp_l.next = ListNode(val%10)
            carry = int(val/10)
            tmp_l = tmp_l.next
            l1 = l1.next
            l2 = l2.next
        if l1:
            tmp_l.next = l1
        elif l2:
            tmp_l.next = l2
        while tmp_l.next:
            val = tmp_l.next.val + carry
            tmp_l.next.val = val % 10
            carry = int(val/10)
            tmp_l = tmp_l.next
        if carry > 0:
            tmp_l.next = ListNode(carry)
        return ret_l.next
        
